### PR TITLE
Emit trace information through kafka

### PR DESCRIPTION
**Pull request by abhinigam**

### CODE REVIEW

File: linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala [hunk 1]
```diff
@@ -116,29 +116,29 @@ object Headers {
     object Deadline {
       val Key = Prefix + "deadline"
 
-      def read(v: String): FDeadline = {
+      def read(v: String): context.Deadline = {
```
Comment (erdody) [inline linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala]: Why was this changed?

File: project/Deps.scala [hunk 2]
```diff
@@ -12,7 +12,7 @@ object Deps {
 
   // networking
   def finagle(mod: String) =
-    "com.twitter" %% s"finagle-$mod" % "6.36.0"
+    "com.twitter" %% s"finagle-$mod" % "6.37.0"
```
Comment (erdody) [inline project/Deps.scala:15]: What do we need from 37?

File: project/Deps.scala [hunk 3]
```diff
@@ -46,4 +46,6 @@ object Deps {
 
   // curator service discovery
   val curatorSD = "org.apache.curator" % "curator-x-discovery" % "2.11.0"
+  // kafka
+  val kafka = "org.apache.kafka" % "kafka_2.10" % "0.8.2.1"
```
Comment (erdody) [inline project/Deps.scala]: Can we use a more recent version? 0.10? (and scala 2.11)?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  type Config = KafkaTelemeterConfig
+  override def configId = "io.buoyant.telemetry.KafkaTelemeterConfig"
+  def configClass = classOf[KafkaTelemeterConfig]
+}
+
+case class KafkaTelemeterConfig(brokerList: String, sampleRate: Float, numRetries: Int, metrics: Boolean, tracing: Boolean) extends TelemeterConfig {
+  private val log = Logger.get(getClass)
+  def mk(params: Stack.Params): KafkaTelemeter = {
+    log.info("Broker list is %s", brokerList)
+    log.info("Number of retries per request is %s", brokerList)
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Doesn't seem to be accurate

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    new KafkaTelemeter("zipkin", numRetries, sampleRate, brokerList, metrics, tracing)
+  }
+}
+
+//Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Change to javadoc?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  //Only support thrift for now.
+  class KafkaRawZipkinTracer(
+    brokerList: String,
+    numRetries: Int = 3, //this is the kafka default.
+    topic: String = "zipkin",
+    timer: Timer = DefaultTimer.twitter,
+    statsReceiver: StatsReceiver = DefaultStatsReceiver
+  ) extends RawZipkinTracer(statsReceiver, timer) {
+    log.info("Initializing kafka zipkin tracer")
+    private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: I would make this a val and move producer creation logic to a separate method (createProducer or so)

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    configs.put(ProducerConfig.RECONNECT_BACKOFF_MS_CONFIG, RECONNECT_BACKOFF_MS_CONFIG.asInstanceOf[AnyRef])
+    configs.put(ProducerConfig.RETRIES_CONFIG, numRetries.asInstanceOf[AnyRef])
+    producer = new KafkaProducer[Array[Byte], Array[Byte]](configs)
+
+    //retries configured from the input.
+    private[this] val scopedReceiver = statsReceiver.scope("log_span")
+    private[this] val okCounter = scopedReceiver.counter("ok")
+    private[this] val errorReceiver = scopedReceiver.scope("error")
+
+    private[this] def serializeBytes(span: Span): Array[Byte] = {
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: serializeSpan?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        log.debug("Sending kafka %i trace events", spans.size)
+        val statEmissionFailures: AtomicBoolean = new AtomicBoolean(false)
+        val semaphore: Semaphore = new Semaphore((spans.size - 1) * -1)
+        spans.foreach {
+          span =>
+            val bytes: Array[Byte] = serializeBytes(span)
+            val record: ProducerRecord[Array[Byte], Array[Byte]] = new ProducerRecord[Array[Byte], Array[Byte]](topic, null, bytes)
+
+            producer.send(record, new Callback() {
+              @Override
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: there's an override keyword

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+      //fails the entire batch is considered to be failed. The batch will only fail if the retries for all the different
+      //stats emitted as part of the batch are exhausted independently of each other
+      // without being able to send it out successfully.
+      // Will try to do early detection. Currently this will enable both
+      //kafka level and higher level retries if the framework supports it.
+      var ex: Throwable = null
+      Future({
+        log.debug("Sending kafka %i trace events", spans.size)
+        val statEmissionFailures: AtomicBoolean = new AtomicBoolean(false)
+        val semaphore: Semaphore = new Semaphore((spans.size - 1) * -1)
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Instead of waiting for the semaphore and passing exception info and so on, can we just wait on the Futures that producer.send returns?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+import org.apache.kafka.clients.producer.ProducerConfig
+import com.twitter.logging.Logger
+
+class KafkaTelemeterInitializer extends TelemeterInitializer {
+  type Config = KafkaTelemeterConfig
+  override def configId = "io.buoyant.telemetry.KafkaTelemeterConfig"
+  def configClass = classOf[KafkaTelemeterConfig]
+}
+
+case class KafkaTelemeterConfig(brokerList: String, sampleRate: Float, numRetries: Int, metrics: Boolean, tracing: Boolean) extends TelemeterConfig {
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: What are the metrics/tracing flags used for?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/TelemeterInitializer.scala [hunk 11]
```diff
@@ -13,7 +13,7 @@ trait TelemeterInitializer extends ConfigInitializer {
 }
 
 @JsonTypeInfo(
-  use = JsonTypeInfo.Id.NAME,
+  use = JsonTypeInfo.Id.CLASS,
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/TelemeterInitializer.scala]: Why is this needed?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+      //stats emitted as part of the batch are exhausted independently of each other
+      // without being able to send it out successfully.
+      // Will try to do early detection. Currently this will enable both
+      //kafka level and higher level retries if the framework supports it.
+      var ex: Throwable = null
+      Future({
+        log.debug("Sending kafka %i trace events", spans.size)
+        val statEmissionFailures: AtomicBoolean = new AtomicBoolean(false)
+        val semaphore: Semaphore = new Semaphore((spans.size - 1) * -1)
+        spans.foreach {
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Let's map first to ProducerRecord and then foreach producer.send

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    numRetries: Int = 3, //this is the kafka default.
+    topic: String = "zipkin",
+    timer: Timer = DefaultTimer.twitter,
+    statsReceiver: StatsReceiver = DefaultStatsReceiver
+  ) extends RawZipkinTracer(statsReceiver, timer) {
+    log.info("Initializing kafka zipkin tracer")
+    private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+    val configs: java.util.Map[String, AnyRef] = new java.util.HashMap[String, AnyRef]
+    configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    configs.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Let's try to avoid class names. In this case, you can pass the serializer objects to the KafkaProducer constructor.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+//Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
+//We will leave linger.ms unchanged for this reason at 1 ms.
+//The reason for this is that the API has been constructed in such a way that a given trace is annotated across multiple
+//different calls rather than gathering up all its annotations and then sending it on its way.
+
+case class KafkaTelemeter(topic: String, numRetries: Int, sampleRate: Float, brokerList: String, metrics: Boolean, tracing: Boolean) extends Telemeter {
+  private val log = Logger.get(getClass)
+  log.info("Initializing kafka telemeter")
+  private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+  private val METADATA_FETCH_TIMEOUT_CONFIG: Int = 3000
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Can we use durations for these (so we make units explicit and avoid mistyping zeroes)?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  private val TIMEOUT_CONFIG: Int = 300
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: If this was intended to be so low, let's add a comment. Also, it seems to be deprecated (replaced by REQUEST_TIMEOUT_MS_CONFIG)

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+  private val MAX_BLOCK_MS: Int = 1000
+  private val RETRY_BACKOFF_MS_CONFIG: Int = 10000
+  private val RECONNECT_BACKOFF_MS_CONFIG: Int = 10000
+  private val CACHE_BYTE_ARRAYS: Int = 100
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Unused?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+  //TODO need to do the tracerCache stuff to prevent duplicates in the config as well as allow support for multiple
+  //kafka clusters. TODO track all the different instances of tracers/dedup them.
+
+  def tracer: Tracer = {
+    val tracerImpl = KafkaRawZipkinTracer.tracerCache.putIfAbsent(
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Can we merge both and use computeIfAbsent instead?

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+      brokerList + topic,
+      new KafkaTracer(new KafkaRawZipkinTracer(brokerList, numRetries, topic), sampleRate)
+    )
+    KafkaRawZipkinTracer.tracerCache.get(brokerList + topic)
+  }
+
+  def cleanup(): Unit = {
+    log.info("Starting cleanup of the kafka telemeter")
+    //This is a stop gap arrangement. Give one minute for all remaining spans to be flushed.
+    Thread.sleep(1 * 60 * 1000)
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Let's extract the timeout constant and use the close(timeout, duration) instead.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+  private val log = Logger.get(getClass)
+  def mk(params: Stack.Params): KafkaTelemeter = {
+    log.info("Broker list is %s", brokerList)
+    log.info("Number of retries per request is %s", brokerList)
+    new KafkaTelemeter("zipkin", numRetries, sampleRate, brokerList, metrics, tracing)
+  }
+}
+
+//Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
+//We will leave linger.ms unchanged for this reason at 1 ms.
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Documentation says the default value is 0ms.

**Review by erdody — CHANGES_REQUESTED** [review]

File: linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala (hunk 1 above)
Comment (AbhiMedallia) [reply linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala]: This is needed because there is a newer version of finagle which is 0.37 which exposes certain tracing classes which are needed.

File: project/Deps.scala (hunk 2 above)
Comment (AbhiMedallia) [reply project/Deps.scala:15]: Please refer above.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 19 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Will set it explicitly to 1 ms.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 14 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: It does not make any sense since the config only takes integer.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 15 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: That is only in the newer versions of kafka. Again look at my previous comments.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 17 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: It seems a little long winded to do so.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 9 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: I think you might have skipped this part of the comments  "  // Will try to do early detection." 
Currently if even a single event in the batch fails then we consider it a failure and exit without waiting for the others to succeed.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/TelemeterInitializer.scala (hunk 11 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/TelemeterInitializer.scala]: I was not able to get it to work with id. We can think about improving this once we have v1 deployed.

File: project/Deps.scala (hunk 3 above)
Comment (AbhiMedallia) [reply project/Deps.scala]: zipkin does not play nice with higher kafka versions. I can switch to scala 2.11 though.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 13 above)
Comment (AbhiMedallia) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: I want to perform minimal changes to the telemeter framework in order to get this to work. This will enable us to get updates from linkerd without having to worry about rebasing issues. (Currently I have made a single line change to an existing linkerd file).

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 18 above)
Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: I am not clear what you mean here. Can you please elaborate?

Comment (AbhiMedallia) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Discussed and confusion was because of kafka version.

File: linkerd/core/src/main/scala/io/buoyant/linkerd/Linker.scala [hunk 20]
```diff
@@ -95,6 +95,7 @@ object Linker {
 
     def mk(defaultTelemeter: Telemeter = NullTelemeter): Linker = {
       // At least one router must be specified
+      println("Initializing a telemeer")
```
Comment (erdody) [inline linkerd/core/src/main/scala/io/buoyant/linkerd/Linker.scala]: Remove or log

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+  override def configId = "io.buoyant.telemetry.KafkaTelemeterConfig"
+  def configClass = classOf[KafkaTelemeterConfig]
+}
+
+case class KafkaTelemeterConfig(brokerList: String, sampleRate: Float, numRetries: Int) extends TelemeterConfig {
+  private val log = Logger.get(getClass)
+  def mk(params: Stack.Params): KafkaTelemeter = {
+    log.info("Broker list is %s", brokerList)
+    log.info("Number of retries per request is %s", numRetries)
+    log.info("Sample rate is" + sampleRate)
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: %s as well

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 22]
```diff
@@ -0,0 +1,214 @@
+package io.buoyant.telemetry
+
+import com.twitter.finagle.stats.StatsReceiver
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Lots of unused imports

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 23]
```diff
@@ -0,0 +1,214 @@
+package io.buoyant.telemetry
```
Comment (erdody) [inline telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Sorry for not catching this earlier, but we need to move this (and the initializer) to a new module (similar to common-metrics or tracelog), core should only have interfaces and base classes.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/TelemeterInitializer.scala (hunk 11 above)
Comment (erdody) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/TelemeterInitializer.scala]: I think you're missing the service descriptor. 
If we create a new module, it should be under resources/META-INF/services/io.buoyant.telemetry.TelemeterInitializer (see tracelog for an example).
Let me know if it doesn't work.

File: telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala (hunk 9 above)
Comment (erdody) [reply telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: You're right. What do you think about this simpler approach:

``` scala
    def sendSpans(spans: Seq[Span]): Future[Unit] = {
      log.debug("Sending kafka %i trace events", spans.size)
      val futures = spans
        .map(span => new ProducerRecord[Array[Byte], Array[Byte]](topic, null, serializeSpan(span)))
        .map(sendRecord)
        .toStream

      Future.join(futures)
    }

    def sendRecord(record: ProducerRecord[Array[Byte], Array[Byte]]): Future[RecordMetadata] = {
      val promise = new Promise[RecordMetadata]
      producer.send(record, new Callback() {
        override def onCompletion(metadata: RecordMetadata, exception: Exception) {
          if (exception != null) {
            val _ = promise.setException(exception)
          } else {
            val _ = promise.setValue(metadata)
          }
        }
      })
      promise
    }
```

File: telemetry/trace-kafka/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+
+import java.util.Map
+import java.util.HashMap
+
+import org.apache.kafka.clients.producer.ProducerConfig
+import com.twitter.logging.Logger
+
+class KafkaTelemeterInitializer extends TelemeterInitializer {
+  type Config = KafkaTelemeterConfig
+  //override def configId = "io.buoyant.telemetry.KafkaTelemeterConfig"
```
Comment (erdody) [inline telemetry/trace-kafka/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Let's remove

File: telemetry/trace-kafka/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+/**
+ * Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
+ * We will leave linger.ms unchanged for this reason at 1 ms.
+ * The reason for this is that the API has been constructed in such a way that a given trace is annotated across multiple
+ * different calls rather than gathering up all its annotations and then sending it on its way.
+ */
+case class KafkaTelemeter(topic: String, numRetries: Int, sampleRate: Float, brokerList: String) extends Telemeter {
+  println("Hello world")
```
Comment (erdody) [inline telemetry/trace-kafka/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Remove?

File: telemetry/trace-kafka/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+  override def configId = "io.l5d.kafkazipkin"
+  def configClass = classOf[KafkaTelemeterConfig]
+}
+
+case class KafkaTelemeterConfig(brokerList: String, sampleRate: Float, numRetries: Int) extends TelemeterConfig {
+  private val log = Logger.get(getClass)
+  def mk(params: Stack.Params): KafkaTelemeter = {
+    log.info("Broker list is %s", brokerList)
+    log.info("Number of retries per request is %s", numRetries)
+    log.info("Sample rate is" + sampleRate)
```
Comment (erdody) [inline telemetry/trace-kafka/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala]: Nit: %s

**Review by erdody — CHANGES_REQUESTED** [review]
2 last minor changes.

**Review by erdody — APPROVED** [review]
